tests/trainer/test_trainer.py

```diff
@@ -1,247 +1,375 @@
-    elif isinstance(instance, AlgorithmHparams):
-        composer_trainer_hparams.algorithms.append(instance)
-    else:
-        pytest.fail(f"Unknown hparams type: {hparams_cls.__name__}")
-    return composer_trainer_hparams.initialize_object()
-
-
-@pytest.mark.parametrize("hparams_cls", _ALL_LOGGERS_CALLBACKS_ALG_PROFILER_HPARAMS)
-def test_fit_on_all_callbacks_loggers_algs_profilers(
-    composer_trainer_hparams: TrainerHparams,
-    dummy_num_classes: int,
-    hparams_cls: Type[hp.Hparams],
-    monkeypatch: pytest.MonkeyPatch,
-    tmpdir: pathlib.Path,
-):
-    trainer = _build_trainer(composer_trainer_hparams, dummy_num_classes, hparams_cls, monkeypatch, tmpdir)
-    trainer.fit()
-
-
-@pytest.mark.parametrize("hparams_cls", _ALL_LOGGERS_CALLBACKS_ALG_PROFILER_HPARAMS)
-def test_multiple_calls_to_fit(
-    composer_trainer_hparams: TrainerHparams,
-    dummy_num_classes: int,
-    hparams_cls: Type[hp.Hparams],
-    monkeypatch: pytest.MonkeyPatch,
-    tmpdir: pathlib.Path,
-):
-    trainer = _build_trainer(composer_trainer_hparams, dummy_num_classes, hparams_cls, monkeypatch, tmpdir)
-    # idempotency test 1: run the FIT event again
-    trainer.engine.run_event(Event.FIT_START)
-    trainer._train_loop()  # using the private method so we don't shutdown the callbacks
-    trainer.state.max_duration = trainer.state.max_duration + trainer.state.max_duration
-    # idempotency test 2: run FIT again. This will trigger another call to FIT_START
-    if issubclass(hparams_cls, LayerFreezingHparams):
-        pytest.xfail("TODO: Layer freezing does not work with a subsequent call to .fit")
-    trainer.fit()  # using the public method so we do shutdown the callbacks
+from torch.utils.data import DataLoader
+
+from composer.algorithms import LayerFreezing
+from composer.algorithms.cutout.cutout import CutOut
+from composer.callbacks import LRMonitor
+from composer.callbacks.run_directory_uploader import RunDirectoryUploader
+from composer.core.callback import Callback
+from composer.core.types import Model
+from composer.loggers import FileLogger, TQDMLogger
+from composer.trainer import Trainer
+from composer.trainer.trainer_hparams import algorithms_registry, callback_registry, logger_registry
+from tests.common import (RandomClassificationDataset, RandomImageDataset, SimpleConvModel, SimpleModel, device,
+                          world_size)
+
+
+class TestTrainerInit():
+
+    @pytest.fixture
+    def config(self):
+        return {
+            'model': SimpleModel(),
+            'train_dataloader': DataLoader(dataset=RandomClassificationDataset()),
+            'eval_dataloader': DataLoader(dataset=RandomClassificationDataset()),
+            'max_duration': '2ep',
+        }
+
+    def test_init(self, config):
+        trainer = Trainer(**config)
+        assert isinstance(trainer, Trainer)
+
+    def test_model_ddp_wrapped(self, config, init_process_group):
+        # dist initialized first with init_process_group fixture
```

Comment:
This fixture should not be required since dist is configured globally here: https://github.com/mosaicml/composer/blob/06ccbb2830581debe1c06451e6c3a277646e2ea2/tests/fixtures/distributed_fixtures.py#L17

Is the test erroring without it?

Reply:
That fixture will be deprecated, so I am testing without that fixture active, and introducing a much more localized one. We can migrate it over if its needed as we continue the refactor.

The issue with the global `dist` configuration is that it inits process groups outside of our `Trainer`. Ideally we want a clean environment for the trainer tests to run in. e.g. if the Trainer's own dist init functions are broken, that won't be caught since this global fixture is working.

Reply:
Could we replace this test and the one below it with one test like the following?

The underlying reason is that I remember there being issues with destroying process groups. (Never traced down the underlying issue, so perhaps it is no longer a thing anymore.) Assuming that is an issue, we don't destory process groups after a test. The first trainer test is the one that initializes the process group, and every other trainer test uses that same progress group (and doesn't attempt to re-initialize it). (Technically, we initialize this with the configure_dist fixture before any trainer tests run, but I don't think that's no longer necessary). Because we don't destroy process groups, we have to run pytest multiple times -- once without distributed and once with rank-one distributed.

Instead, could we condition this test (and the one that follows) on which pytest environment it is in with something like the following?

```
    def test_model_ddp_wrapped(self, config):
        # dist initialized first with init_process_group fixture
        trainer = Trainer(**config)
        should_be_ddp_wrapped = "WORLD_SIZE" in os.environ  # if WORLD_SIZE and all other dist environment variables are set, the the trainer will configure dist. Otherwise, it shouldn't, unless if deepspeed is being used.
        assert isinstance(trainer.state.model, DistributedDataParallel) == should_be_ddp_wrapped
```